packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+        if (routeInformation != null) {
+          _lastSeenLocation = routeInformation.location;
+        }
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+
+      case _IntentionToReportRouteInformation.maybe:
+        final RouteInformation routeInformation = _retrieveNewRouteInformation;
+        if (routeInformation != null) {
+          if (_lastSeenLocation != routeInformation.location) {
+            widget.routeInformationProvider
+              .routerReportsNewRouteInformation(routeInformation);
+            _lastSeenLocation = routeInformation.location;
+          }
+        }
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+
+      case _IntentionToReportRouteInformation.must:
+        final RouteInformation routeInformation = _retrieveNewRouteInformation;
+        if (routeInformation != null) {
+          widget.routeInformationProvider
+            .routerReportsNewRouteInformation(routeInformation);
+          _lastSeenLocation = routeInformation.location;
+        }
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+    }
+  }
+
+  RouteInformation get _retrieveNewRouteInformation {
+    final T configuration = widget.routerDelegate.currentConfiguration;
+    if (configuration == null)
+      return null;
+    final RouteInformation routeInformation = widget
+      .routeInformationParser
+      .restoreRouteInformation(configuration);
+    assert((){
+      if (routeInformation == null) {
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Router.routeInformationParser returns a null RouteInformation. '
+              'If you opt for route information reporting, the '
+              'routeInformationParser must not report null for a given '
+              'configuration.'
+          ),
+        );
+      }
+      return true;
+    }());
+    return routeInformation;
+  }
+
+  void _setStateWithExplicitReportStatus(
+    _IntentionToReportRouteInformation status,
+    VoidCallback fn,
+  ) {
+    assert(status != null);
+    assert(status.index >= _IntentionToReportRouteInformation.must.index);
+    assert(() {
+      if (_currentIntentionToReporting.index >= _IntentionToReportRouteInformation.must.index &&
+          _currentIntentionToReporting != status) {
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Both Router.navigate and Router.neglect have been called in this '
+              'build cycle, and the Router cannot decide whether to report the '
+              'route information. Please make sure only one of them is called '
+              'within the same build cycle.'
+          ),
+        );
+      }
+      return true;
+    }());
+    _currentIntentionToReporting = status;
+    setState(fn);
```

Comment:
Technically, this is an unnecessary rebuild. Just because somebody wants to add a new history entry for whatever reason, doesn't mean we need to rebuild the router, right? If we uncouple the reporting from buidling, e.g. with a postframecallback we could avoid that, no?